libphal/phal_pdbg.C

```diff
@@ -300,24 +303,149 @@ bool isSbeVitalAttnActive(struct pdbg_target *proc)
+
+	// Is there any error in reading the scom address, consider control is
+	// in hostboot
+	if (pib_read(pibTarget, l_coreScratchRegAddr, &l_coreScratchRegData)) {
+		// If unable to read the register, by default consider the
+		// control is in hostboot
+		log(level::ERROR, "scom read error: 0x%X",
+		    l_coreScratchRegAddr);
+		return false;
+	}
+
+	// If the register reads zero, return control moved to host.
+	return (l_coreScratchRegData == 0);
+}
+
+std::optional<LocationCode> getUnexpandedLocCode(const LocationCode &locCode)
+{
+	constexpr uint8_t EXP_LOCATION_CODE_MIN_LENGTH = 17;
+	// Location code should start with "U"
+	if (locCode[0] != 'U') {
+		log(level::ERROR,
+		    "Location code should start with \"U\""
+		    " but, given location code %s",
+		    locCode.c_str());
+		return std::nullopt;
+	}
+
+	// Given location code length should be need to match with minimum
+	// length
+	// to drop expanded format in given location code.
+	if (locCode.length() < EXP_LOCATION_CODE_MIN_LENGTH) {
+		log(level::ERROR,
+		    "Given location code %s is not meet "
+		    "with minimum length %d",
+		    locCode.c_str(), EXP_LOCATION_CODE_MIN_LENGTH);
+		return std::nullopt;
+	}
+
+	// "-" should be there to seggregate all (FC, Node number and SE values)
+	// together.
+	// Note: Each (FC, Node number and SE) value can be seggregate by "."
+	// but, cec device tree just have unexpand format so, just skipping
+	// still first occurrence "-" and replacing with "fcs".
+	auto endPosOfFcs = locCode.find('-', EXP_LOCATION_CODE_MIN_LENGTH);
+	if (endPosOfFcs == std::string::npos) {
+		log(level::ERROR,
+		    "Given location code %s is not valid i.e could not find "
+		    "dash",
+		    locCode.c_str());
+		return std::nullopt;
+	}
+
+	std::string unExpandedLocCode("Ufcs");
+
+	if (locCode.length() > EXP_LOCATION_CODE_MIN_LENGTH) {
+		unExpandedLocCode.append(
+		    locCode.substr(endPosOfFcs, std::string::npos));
+	}
+
+	return unExpandedLocCode;
+}
+
+/**
+ * @brief Fetch FRU type from device tree.
+ *
+ * @details An api to get FRU ATTR_TYPE corresponding to
+ *          a given unexpanded location code.
+ *
+ * @param[in] locationCode - Location code in unexpanded format.
+ * @param[out] fruType - Fru type of a given location code.
+ *
+ * @return 0 if fru type is found.
+ * @return -1 if unable to find a fru with the given location code.
+ * @return DEVTREE_ATTR_READ_FAIL if unable to fetch ATTR_TYPE from device tree.
+ */
+static int fetchFruTypeFromDevTree(const LocationCode &locationCode,
+				   ATTR_TYPE_Type &fruType)
+{
+	ATTR_LOCATION_CODE_Type locCode;
+	// Defining a list of existing frus
+	// Note: We need to update the list if a new fru is added
+	std::vector<const char *> frus{"proc", "tpm", "dimm"};
+	struct pdbg_target *target;
+
+	for (auto fru : frus) {
+		pdbg_for_each_class_target(fru, target)
+		{
+			if (DT_GET_PROP(ATTR_LOCATION_CODE, target, locCode)) {
+				continue;
+			} else {
```

Comment:
please make this else-if and remove the nested if condition.

Reply:
return 0 can independently exist after the inner if condition, that inner else can also be removed

Reply:
else if (locCode == locationCode)
 {
	if (DT_GET_PROP(ATTR_TYPE, target, fruType)) {
		return openpower::phal::exception::DEVTREE_ATTR_READ_FAIL;
	}
        return 0;
}